### PR TITLE
Classic average function.

### PR DESCRIPTION
**Pull request by cooffeeRequired**
### Description
Classic average function.

### Example
```K
[00:25:58 INFO]: executing 'send average(10,190,299.20)'
[00:25:58 INFO]: 166.4
```

---
**Target Minecraft:** **N/A**
**Requirements: ** **N/A**
**Related Issues:** **N/A**

### CODE REVIEW

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 1]
```diff
@@ -56,6 +56,27 @@ private static String str(double n) {
 		Parameter<?>[] numbersParam = new Parameter[] {new Parameter<>("ns", DefaultClasses.NUMBER, false, null)};
 		
 		// basic math functions
+
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:59]:
```suggestion
```

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 2]
```diff
@@ -56,6 +56,27 @@ private static String str(double n) {
...
+				Double sum = 0.0;
+				int length = params[0].length;
+				for (int i = 0; i < length; i++) {
+					if (params[0][i] instanceof Double) {
+						sum += (Double) params[0][i];
+					} else if (params[0][i] instanceof Integer) {
+						sum += ((Integer) params[0][i]).doubleValue();
+					} else if (params[0][i] instanceof Long) {
+						sum += ((Long) params[0][i]).doubleValue();
+					}
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:73]:
```suggestion
					if (params[0][i] instanceof Number)
						sum += ((Number) params[0][i]).doubleValue();
```

**Review by TheLimeGlass — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 3]
```diff
@@ -56,6 +56,21 @@ private static String str(double n) {
 		Parameter<?>[] numbersParam = new Parameter[] {new Parameter<>("ns", DefaultClasses.NUMBER, false, null)};
 		
 		// basic math functions
+
+		Functions.registerFunction(new SimpleJavaFunction<Number>("average", new Parameter[] {new Parameter<>("list", DefaultClasses.NUMBER, false, null)}, DefaultClasses.NUMBER, true) {
+			@Override
+			public Number[] executeSimple(Object[][] params) {
+				Double sum = 0.0;
```
Comment (UnderscoreTud) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:63]:
```suggestion
				double sum = 0.0;
```

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 4]
```diff
@@ -56,6 +56,21 @@ private static String str(double n) {
 		Parameter<?>[] numbersParam = new Parameter[] {new Parameter<>("ns", DefaultClasses.NUMBER, false, null)};
 		
 		// basic math functions
+
+		Functions.registerFunction(new SimpleJavaFunction<Number>("average", new Parameter[] {new Parameter<>("list", DefaultClasses.NUMBER, false, null)}, DefaultClasses.NUMBER, true) {
```
Comment (UnderscoreTud) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:60]:
```suggestion
		Functions.registerFunction(new SimpleJavaFunction<Number>("average", numbersParam, DefaultClasses.NUMBER, true) {
```
There's already a static field for this type of parameter

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 5]
```diff
@@ -56,6 +56,21 @@ private static String str(double n) {
...
+			public Number[] executeSimple(Object[][] params) {
+				Double sum = 0.0;
+				int length = params[0].length;
+				for (int i = 0; i < length; i++) {
+					if (params[0][i] instanceof Number)
+						sum += ((Number) params[0][i]).doubleValue();
+				}
+				return new Number[] {sum / length};
+			}
+		}.description("Returns the average of all numbers in the list")
```
Comment (UnderscoreTud) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:71]:
```suggestion
		}.description("Returns the average of all numbers in the list.")
```

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 6]
```diff
@@ -56,6 +56,21 @@ private static String str(double n) {
...
 		
 		// basic math functions
+
+		Functions.registerFunction(new SimpleJavaFunction<Number>("average", new Parameter[] {new Parameter<>("list", DefaultClasses.NUMBER, false, null)}, DefaultClasses.NUMBER, true) {
+			@Override
+			public Number[] executeSimple(Object[][] params) {
+				Double sum = 0.0;
+				int length = params[0].length;
+				for (int i = 0; i < length; i++) {
+					if (params[0][i] instanceof Number)
```
Comment (UnderscoreTud) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:66]: No need to check if it's an instance of `Number`

**Review by UnderscoreTud — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java (hunk 6 above)
Comment (cooffeeRequired) [reply src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:66]: Added by Pickle :D

Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:66]: It's an appropriate change. We talked over it. Skript already ensures the Object is of the appropriate type and that it's not null.

Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:66]: @cooffeeRequired unresolved

**Review by TheLimeGlass — CHANGES_REQUESTED** [review]
Some tests could be implemented for this.

**Comment by cooffeeRequired** [comment]
@TheLimeGlass done

**Review by TheLimeGlass — APPROVED** [review]

**Review by TheLimeGlass — CHANGES_REQUESTED** [review]
Tests failing

**Comment by TheLimeGlass** [comment]
use `gradlew clean quickTest` to test

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 7]
```diff
@@ -56,6 +56,20 @@ private static String str(double n) {
...
+			public Number[] executeSimple(Object[][] params) {
+				double sum = 0.0;
+				int length = params[0].length;
+				for (int i = 0; i < length; i++) {
+					sum += ((Number) params[0][i]).doubleValue();
+				}
+				return new Number[] {sum / length};
+			}
+		}.description("Returns the average of all numbers in the list.")
+			.examples("average(10, 20) = 15", "average(11, 99) = 41", "average(10, 190, 299.20) = 166.2")
```
Comment (TPGamesNL) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:71]:
```suggestion
			.examples("average(10, 20) = 15", "average(11, 99) = 55", "average(10, 190, 299.20) = 166.4")
```

File: src/test/skript/tests/syntaxes/functions/Average.sk [hunk 8]
```diff
@@ -0,0 +1,4 @@
+test "average function":
+	assert average(10, 300, 300, 3000) = 902.5 with "average function passed ##1"
+	assert average(10, 3002, -320, -300) = 598 with "average function passed ##2"
+	assert average(400, 20, 1, 328.2, 39.22) = 157.68 with "average function passed ##3"
```
Comment (TPGamesNL) [inline src/test/skript/tests/syntaxes/functions/Average.sk:4]:
```suggestion
	assert average(400, 20, 1, 328.2, 39.22) = 157.684 with "average function passed ##3"

```

**Review by TPGamesNL — CHANGES_REQUESTED** [review]

**Comment by cooffeeRequired** [comment]
why did it fail?

**Review by TheLimeGlass — APPROVED** [review]

**Review by AyhamAl-Ali — APPROVED** [review]

**Review by UnderscoreTud — APPROVED** [review]

**Pull request closed without merging**